_scripts/libdocker.sh

```diff
@@ -0,0 +1,144 @@
+export DKR_REGISTRY=${DKR_REGISTRY:-https://registry.hub.docker.com}
+export DKR_REGISTRY_VERSION=${DKR_REGISTRY_VERSION:-v2}
+EOF
+}
+
+########################
+# Login to target registry.
+# Arguments:
+#   None
+# Returns:
+#   None
+dkr_login() {
+    local -r username="${DKR_USERNAME:?docker username required}"
+    local -r password="${DKR_PASSWORD:?docker password required}"
+    local -r server="${DKR_SERVER:?docker server required}"
+
+    echo "${password}" | dkr_execute login -u "${username}" --password-stdin "${server}"
+}
+
+########################
+# Check if docker image and tag exists in target registry.
+# Arguments:
+#   $1 - Docker image name
+#   $2 - Docker image tag (Optional. Default: latest)
+# Returns:
+#   Boolean
+dkr_image_exists() {
+    local -r image_name="${1:?docker image name argument required}"
+    local -r tag="${2:-latest}"
+    local -r url="$DKR_REGISTRY/$DKR_REGISTRY_VERSION/repositories/$DKR_USERNAME/$image_name/tags/$tag"
+
+    http_response=$(curl_execute "$url")
+    # extract the status
+    http_status=$(echo "$http_response" | tr -d '\n' | sed -e 's/.*HTTPSTATUS://')
+
+    if [ "$http_status" -eq 200 ]; then
+        echo true
+        return
+    fi
+    echo false
+}
+
+########################
+# Build Dockerfile image.
+# Arguments:
+#   $1 - Build context path
+#   $2 - Docker image name
+# Returns:
+#   None
+dkr_build() {
+    local -r context="${1:?docker build context path is required}"
+    local -r image_name="${2:?docker image name argument required}"
+
+    if [ ! -d "$context" ]; then
+        echo "Context path provided [$context] does not exist or is not a directory" 1>&2
+        exit 1
+    fi
+    cd "$context" || exit
+    dkr_execute build --quiet -t "$DKR_USERNAME/$image_name" . > /dev/null
+    cd - > /dev/null || return
+}
+
+########################
+# Add tags to docker image.
+# Arguments:
+#   $1 - Docker image name
+#   $@ - List of tags to apply to the image
+# Returns:
+#   None
+dkr_tag() {
+    local -r image_name="${1:?docker image name argument required}"
+    shift
+    read -r -a tags <<< "$(tr ',;' ' ' <<< "$@")"
+
+    if ! dkr_execute image inspect "$DKR_USERNAME/$image_name" > /dev/null; then
+        echo "Docker image [$DKR_USERNAME/$image_name] does not exist" 1>&2
+        exit 1
+    fi
+
+    for t in "${tags[@]}"; do
+        dkr_execute tag "$DKR_USERNAME/$image_name" "$DKR_USERNAME/$image_name:$t"
+    done
+}
+
+########################
+# Push docker image to registry.
+# Arguments:
+#   $1 - Docker image name
+# Returns:
+#   String
+dkr_push() {
+    local -r image_name="${1:?docker image name argument required}"
+
+    if ! dkr_execute image inspect "$DKR_USERNAME/$image_name" > /dev/null; then
+        echo "Docker image [$DKR_USERNAME/$image_name] does not exist" 1>&2
+        exit 1
+    fi
+    # Ensure we are logged in
+    dkr_login
+    dkr_execute push "$DKR_USERNAME/$image_name" > /dev/null
```

Comment:
Add $DKR_REGISTRY to use the exact registry.

Reply:
I'm having troubles with this one.
I will skip for now but it shouldn't be a problem if is properly defined in your local configuration.

Reply:
Yes, we can keep that for another iteration